### PR TITLE
Restrict istio-cni-node DaemonSet

### PR DESCRIPTION
**Pull request by InnocentK**
**Please provide a description of this PR:**
istio-cni by default is installed on all nodes. This PR adds the ability to exclude istio-cni from being installed on specified nodes by adding the istio.io/exclude-cni label to that node and setting it to true which updates the istio-cni nodeSelector daemon set.

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
Do we want opt-in (allow users to add a NodeSelector) or opt-out?

**Review by howardjohn — COMMENTED** [review]
This doesn't allow adding the label to the node after CNI starts running as far as I can tell

**Review by Arhell — COMMENTED** [review]
/ok-to-test

**Comment by zirain** [comment]
curious why `gen-check` pass

**Review by jewertow — COMMENTED** [review]
@InnocentK could you add a release note? I think that "installation" is the right area for this change.

**Comment by jewertow** [comment]
> This doesn't allow adding the label to the node after CNI starts running as far as I can tell

@howardjohn could you tell more what you mean? How CNI would block adding a label? Or did you mean that setting a label after a CNI container has been scheduled does not remove that container?

**Comment by howardjohn** [comment]
> > This doesn't allow adding the label to the node after CNI starts running as far as I can tell
> 
> @howardjohn could you tell more what you mean? How CNI would block adding a label? Or did you mean that setting a label after a CNI container has been scheduled does not remove that container?

It does not remove the container, unlike nodeSelector would

**Comment by dgn** [comment]
> This doesn't allow adding the label to the node after CNI starts running as far as I can tell

That's true, but I think opt-out is preferable here, and then your only option really is affinity, so I think this solution is fine. If we go this way, we should document that you need to `kubectl rollout restart daemonset` your istio-cni daemonset after adding the label to nodes

**Comment by howardjohn** [comment]
Can you explain the use case for opt-out instead of opt-in? I have only seen daemonsets use node selector previously. 

It definitely would need to be a setting for `nodeSelector`, where users add an explicit label (and then assign some nodes that label).

I am not terribly against the affinity, it has some clear advantages over the nodeSelector. I had just not previously seen it done before so my initial reaction was that it seemed odd. We could have both in theory as well, they are not exclusive

**Comment by InnocentK** [comment]
> Can you explain the use case for opt-out instead of opt-in? I have only seen daemonsets use node selector previously.
> 
> It definitely would need to be a setting for `nodeSelector`, where users add an explicit label (and then assign some nodes that label).
> 
> I am not terribly against the affinity, it has some clear advantages over the nodeSelector. I had just not previously seen it done before so my initial reaction was that it seemed odd. We could have both in theory as well, they are not exclusive

The reason I'm preferring to use opt-out in this case is to maintain the default use of istio-cni. Since it has been available for some time now, the average user likely assumes its there by default and I wanted to get away from disrupting this. I think there is a greater number of use cases for wanting istio-cni installed than removed and I wanted that reflected

**Comment by InnocentK** [comment]
> curious why `gen-check` pass

I don't think istio-cni is affected so there were no changes this time.

File: releasenotes/notes/restrict-istio-cni-node-daemonset.yaml
Comment (InnocentK) [reply releasenotes/notes/restrict-istio-cni-node-daemonset.yaml:6]: This is much clearer. Thanks!

**Review by dhawton — APPROVED** [review]
Release note looks good.

**Review by zirain — CHANGES_REQUESTED** [review]
IMO, make `affinity` configurable in `values.yaml` is better than use static `istio.io/exclude-cni`

**Comment by InnocentK** [comment]
> IMO, make `affinity` configurable in `values.yaml` is better than use static `istio.io/exclude-cni`

Are you suggesting that the key checked in the affinity be set in values.yaml?

**Comment by zirain** [comment]
prefer to something like https://github.com/istio/istio/blob/916936a123afa8d393d3b5d9ee623dabe1314361/manifests/charts/gateway/values.yaml#L88

File: manifests/charts/istio-cni/templates/daemonset.yaml
Comment (zirain) [inline manifests/charts/istio-cni/templates/daemonset.yaml:48]: I don't think L48-L55 is needed

File: manifests/charts/istio-cni/templates/daemonset.yaml
Comment (zirain) [inline manifests/charts/istio-cni/templates/daemonset.yaml:47]: move this under the if in L48?

**Comment by InnocentK** [comment]
@zirain Thanks! I was able to get the changes you suggested working now

File: releasenotes/notes/restrict-istio-cni-node-daemonset.yaml
Comment (linsun) [inline releasenotes/notes/restrict-istio-cni-node-daemonset.yaml:6]: What are the scenarios of why someone would want to do it?

Comment (linsun) [reply releasenotes/notes/restrict-istio-cni-node-daemonset.yaml:6]: What is the experience with ambient? If you label a namespace but it lands on a pod that don't have istio-cni....

Comment (howardjohn) [reply releasenotes/notes/restrict-istio-cni-node-daemonset.yaml:6]: Same applies to sidecars

Comment (jewertow) [reply releasenotes/notes/restrict-istio-cni-node-daemonset.yaml:6]: The main use case is multi-tenancy. When users have big k8s clusters, they may want to install service mesh only for a subset of services, because:
1. Service mesh fits only for a minor part of their workloads.
2. Some of their workloads require special care and run on dedicated nodes that are not part of service mesh.
3. They deploy production and test workloads in the same cluster and use node selectors to manage these environments.

Comment (jewertow) [reply releasenotes/notes/restrict-istio-cni-node-daemonset.yaml:6]: > What is the experience with ambient?

I guess we will need a similar or the same option in ztunnel daemon set. Do you expect it to be implemented while it's still in experimental phase?

> Same applies to sidecars

This is optional setting, so I assume that users should use it consciously. Do you expect a mechanism to prevent injecting sidecars to workloads on nodes without CNI pods? In such a case, injecting would fail silently, as it would not execute at all. In case of missing CNI node, injected sidecar should fail at startup, and this seems better to me.

**Review by zirain — COMMENTED** [review]
lgtm, defer to other maintainers

**Comment by InnocentK** [comment]
> lgtm, defer to other maintainers

@zirain Any other reviews would be blocked by your previous "requested changes" review. Would you mind updating that? Thanks in advance

**Review by zirain — DISMISSED** [review]

**Comment by InnocentK** [comment]
@jacob-delgado @costinm @Monkeyanator Could I get a review from you? Thank you in advance.

**Comment by dgn** [comment]
not stale

**Comment by InnocentK** [comment]
Not stale. Still needs reviews from @jacob-delgado @costinm @Monkeyanator

**Review by howardjohn — APPROVED** [review]

**Review by jacob-delgado — APPROVED** [review]

**Pull request merged**
